### PR TITLE
RD-2809: Integration tests provided for the "pluginsNum" widget 

### PR DESCRIPTION
**Pull request by jawamartus**
## Description
All the cases mentioned in the JIRA ticket has been covered.

## Screenshots / Videos
N/A

## Checklist

- [x] My code follows the UI Code Style.
- [x] I verified that all tests and checks have passed.
- [x] I verified if that change requires a documentation update.
- [x] I added proper labels to this PR.

## Tests

I verified cypress changed tests locally. All of them passed.

All tests passed: https://jenkins.cloudify.co/blue/organizations/jenkins/Stage-UI-System-Test/detail/Stage-UI-System-Test/1875/pipeline

 

## Documentation
N/A

### CODE REVIEW

File: test/cypress/integration/widgets/plugins_num_spec.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
+describe('Number of Plugins widget', () => {
+    const widgetId = 'pluginsNum';
+
+    function clickOnWidget() {
+        cy.get('.pluginsNumWidget .statistic').click();
+    }
+
+    function verifyUrl(expectedPageId: string) {
+        cy.location('pathname').should('be.equal', `/console/page/${expectedPageId}`);
+    }
```
Comment (qooban) [inline test/cypress/integration/widgets/plugins_num_spec.ts:10]: Seeing that the same function is defined in `blueprints_num_spec.ts` maybe worth to extract it to custom Cypress command? Probably the best fit would be `test/cypress/support/commands.ts` (it seems like we should cut it to smaller pieces and have few files containing different groups of those custom commands)

The name could be `verifyPageUrl` I think.

File: test/cypress/integration/widgets/plugins_num_spec.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        cy.setSearchableDropdownConfigurationField(widgetId, 'Page to open on click', pageId);
+        clickOnWidget();
+
+        verifyUrl(pageId);
+    });
+
+    it('displays the correct number of plugins', () => {
+        cy.stageRequest('/console/sp/plugins', 'GET').then(data => {
+            const num = _.get(data.body, 'items', []).length;
+            cy.addWidget(widgetId)
```
Comment (qooban) [inline test/cypress/integration/widgets/plugins_num_spec.ts:35]: Do we need another widget in the page?

File: widgets/pluginsNum/src/widget.tsx [hunk 3]
```diff
@@ -23,7 +23,7 @@ Stage.defineWidget({
             name: 'Page to open on click',
             description: 'Page to open when user clicks on widget content',
             type: Stage.Basic.GenericField.CUSTOM_TYPE,
-            default: 'system_resources',
+            default: 'plugins',
```
Comment (qooban) [inline widgets/pluginsNum/src/widget.tsx:26]: Good catch! You've just fixed a bug 🙂 

With proper test for checking if default page is available in default templates we'll have better verification on such cases.

File: test/cypress/integration/widgets/plugins_num_spec.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
+describe('Number of Plugins widget', () => {
+    const widgetId = 'pluginsNum';
+
+    function clickOnWidget() {
+        cy.get('.pluginsNumWidget .statistic').click();
```
Comment (qooban) [inline test/cypress/integration/widgets/plugins_num_spec.ts:5]: Maybe:
```suggestion
        cy.getWidget(widgetId).find('.statistic').click();
```
would be a bit cleaner?

File: test/cypress/integration/widgets/plugins_num_spec.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        clickOnWidget();
+
+        verifyUrl(pageId);
+    });
+
+    it('displays the correct number of plugins', () => {
+        cy.stageRequest('/console/sp/plugins', 'GET').then(data => {
+            const num = _.get(data.body, 'items', []).length;
+            cy.addWidget(widgetId)
+                .get('.pluginsNumWidget .value')
```
Comment (qooban) [inline test/cypress/integration/widgets/plugins_num_spec.ts:36]: What do you think about having:
```suggestion
                .getWidget(widgetId).find('.value')
```
?

**Review by qooban — COMMENTED** [review]
Added some comments, mostly minor.

Try to fill in PR details (according to description provided in commented parts).

During tomorrow meeting we can discuss PR description, testing, documentation and all PR related stuff.

File: test/cypress/integration/widgets/plugins_num_spec.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+    function clickOnWidget() {
+        cy.get('.pluginsNumWidget .statistic').click();
+    }
+
+    function verifyUrl(expectedPageId: string) {
+        cy.location('pathname').should('be.equal', `/console/page/${expectedPageId}`);
+    }
+
+    before(() => cy.activate('valid_trial_license').mockLogin().addWidget(widgetId));
```
Comment (qooban) [inline test/cypress/integration/widgets/plugins_num_spec.ts:12]: You can try sth like:
```
    before(() => cy.activate('valid_trial_license').mockLogin().addPage('test').addWidget(widgetId));

    beforeEach(() => cy.visitPage('test'));
```

Comment (jawamartus) [reply test/cypress/integration/widgets/plugins_num_spec.ts]: Done!

File: test/cypress/integration/widgets/plugins_num_spec.ts (hunk 5 above)
Comment (jawamartus) [reply test/cypress/integration/widgets/plugins_num_spec.ts]: Done!

File: test/cypress/integration/widgets/plugins_num_spec.ts (hunk 4 above)
Comment (jawamartus) [reply test/cypress/integration/widgets/plugins_num_spec.ts]: Done!

File: test/cypress/integration/widgets/plugins_num_spec.ts (hunk 2 above)
Comment (jawamartus) [reply test/cypress/integration/widgets/plugins_num_spec.ts]: Removed! Thanks!

File: test/cypress/integration/widgets/plugins_num_spec.ts (hunk 1 above)
Comment (jawamartus) [reply test/cypress/integration/widgets/plugins_num_spec.ts]: Done!

File: test/cypress/integration/widgets/plugins_num_spec.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+    it('Opens the default page on click', () => {
+        const defaultPage = 'plugins';
+
+        clickOnWidget();
+        verifyUrl(defaultPage);
+    });
```
[the remark this answers is not in the document]
Comment (jawamartus) [reply test/cypress/integration/widgets/plugins_num_spec.ts]: Done!

File: test/cypress/support/commands.ts [hunk 8]
```diff
@@ -446,7 +447,8 @@ const commands = {
     mockDisabledGettingStarted: () => mockGettingStarted(false),
 
     getWidget: (widgetId: string) => cy.get(`.${widgetId}Widget`),
-
+    verifyLocationByPageId: (expectedPageId: string) =>
+        cy.location('pathname').should('be.equal', `/console/page/${expectedPageId}`),
```
Comment (qooban) [inline test/cypress/support/commands.ts:451]: Actually I found better place for that command. Check `test/cypress/support/asserts.ts`.

File: test/cypress/support/commands.ts [hunk 9]
```diff
@@ -213,11 +213,12 @@ const commands = {
         });
         return cy.visit('/console');
     },
+    // Note: The following function shouldn't be responsible for location verification
```
Comment (qooban) [inline test/cypress/support/commands.ts:216]: Hmm... I'm not sure, maybe the intention of author was to wait for the URL to change? I'm not sure it is valid comment.

**Review by qooban — COMMENTED** [review]
Looks better IMHO.

1. Seeing unticked Checklist I'm wondering should we clarify something together? I won't have time today I think, sorry. If necessary, let's discuss on Monday.
2. Reg.
> I verified cypress changed tests locally. All of them passed.

I think it is not enough. Running locally vs. running in Jenkins it's not the same. Please trigger full test run. It costs us almost nothing (you just need to wait for the results), but saves time in the future.

File: test/cypress/support/commands.ts (hunk 8 above)
Comment (jawamartus) [reply test/cypress/support/commands.ts]: Good point. Moved.

**Review by qooban — APPROVED** [review]
Looks good! 👍🏻 

I assume all the items in Checklist can be ticked off? (see my comment from the previous review https://github.com/cloudify-cosmo/cloudify-stage/pull/1762#pullrequestreview-859487620 - point 1)

**Pull request merged**